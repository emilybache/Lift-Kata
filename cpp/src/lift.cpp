//
// Created by Anders Arnholm on 2019-12-14.
//

#include <sstream>
#include <iomanip>
#include "lift.h"

std::string print_lifts(const LiftSystem &system) {
    std::stringstream buffer;
    size_t total_calls = system._calls.size();
    for (auto floor : system._floors) {
        std::set<Call> calls_for_floor;
        buffer << std::setw(3) << floor << std::setw(0) << " ";
        std::copy_if(system._calls.begin(),
                     system._calls.end(),
                     std::inserter(calls_for_floor, calls_for_floor.end()),
                     [floor](const Call &call) { return call.floor == floor; }
        );
        std::stringstream call_buffer;
        for (auto call : calls_for_floor) {
            call_buffer << (call.direction == Call::Down ? "v" : "^");
        }
        buffer << std::setw(total_calls) << call_buffer.str();
        for (const auto &lift : system._lifts) {
            std::string request_identifier;
            if (lift._requested_floors.count(floor)) {
                request_identifier = "*";
            }
            std::string lift_label;
            if (lift._floor == floor) {
                if (lift._doors_open) {
                    lift_label = "]" + request_identifier + lift._id + "[";
                } else {
                    lift_label = "[" + request_identifier + lift._id + "]";
                }
            } else {
                lift_label = request_identifier + "  ";
            }
            size_t lift_label_width = 5 + lift._id.length();
            buffer << std::setw(lift_label_width) << lift_label << std::setw(0);
        }
        buffer << std::setw(0) << "  " << std::setw(3) << floor << std::endl;
    }
    return buffer.str();
}
