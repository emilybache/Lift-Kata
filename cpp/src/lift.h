//
// Created by Anders Arnholm on 2019-12-14.
//

#ifndef CPP_LIFT_H
#define CPP_LIFT_H

#include <string>
#include <utility>
#include <vector>
#include <set>

class Call {
public:
    bool operator<(const Call &rhs) const { // Needed for std::inserter()
        if (floor < rhs.floor)
            return true;
        if (rhs.floor < floor)
            return false;
        return direction < rhs.direction;
    }

    int floor;
    enum Direction {
        Up, Down
    } direction;

    Call(int floor, Direction direction) : floor(floor), direction(direction) {}
};

class LiftSystem;

class Lift {
private:
    std::string _id;
    int _floor;
    bool _doors_open;
    std::set<int> _requested_floors;
public:
    Lift(std::string id, int floor, bool doors_open, std::set<int> requested_floors)
            : _id(std::move(id)), _floor(floor), _doors_open(doors_open),
              _requested_floors(std::move(requested_floors)) {}

    friend std::string print_lifts(const LiftSystem &system);
};

class LiftSystem {
private:
    std::vector<Lift> _lifts;
    std::vector<int> _floors;
    std::vector<Call> _calls;
public:
    LiftSystem(std::vector<Lift> lifts,
               std::vector<int> floors,
               std::vector<Call> calls) :
            _lifts(std::move(lifts)),
            _floors(std::move(floors)),
            _calls(std::move(calls)) {}

    void tick() {
        // TODO For the user to implement
    }

    friend std::string print_lifts(const LiftSystem &system);
};

std::string print_lifts(const LiftSystem &system);

#endif //CPP_LIFT_H
