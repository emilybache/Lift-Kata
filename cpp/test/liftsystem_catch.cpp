//
// Created by Anders Arnholm on 2019-12-27.
//

#include "catch2/catch.hpp"
#include "ApprovalTests.hpp"

#include "lift.h"

TEST_CASE("EmptySystem") {
    auto lift_system = LiftSystem(std::vector<Lift>(),
                                  std::vector<int>{0, 1, 2, 3}, std::vector<Call>());
    ApprovalTests::Approvals::verify(print_lifts(lift_system));
}

TEST_CASE("SampleLiftSystem") {
    const std::vector<int> &floors = std::vector<int>{0, 1, 2, 3};
    const std::vector<Lift> &lifts = std::vector<Lift>{
            Lift("A", 3, false, std::set<int>{0}),
            Lift("B", 2, false, std::set<int>{}),
            Lift("C", 2, true, std::set<int>{}),
            Lift("D", 0, false, std::set<int>{0})
    };
    const std::vector<Call> &calls = std::vector<Call>{Call(1, Call::Down)};
    auto lift_system = LiftSystem(lifts, floors, calls);
    ApprovalTests::Approvals::verify(print_lifts(lift_system));
}

TEST_CASE("IllegalState") {
    auto lift_system = LiftSystem(std::vector<Lift>{Lift("A", 0, true, std::set<int>{0})},
                                  std::vector<int>{0, 1}, std::vector<Call>());
    ApprovalTests::Approvals::verify(print_lifts(lift_system));
}

TEST_CASE("LargeLiftSystem") {
    const std::vector<int> &floors = std::vector<int>{-2, -1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
    const std::vector<Lift> &lifts = std::vector<Lift>{
            Lift("A", 0, false, std::set<int>{3, 5, 7}),
            Lift("B", 2, true, std::set<int>{}),
            Lift("C", -2, false, std::set<int>{-2, 0}),
            Lift("D", 8, true, std::set<int>{0, -1, -2}),
            Lift("SVC", 10, true, std::set<int>{0, -1}),
            Lift("F", 8, true, std::set<int>{}),
    };
    const std::vector<Call> &calls = std::vector<Call>{Call(1, Call::Down)};
    auto lift_system = LiftSystem(lifts, floors, calls);
    ApprovalTests::Approvals::verify(print_lifts(lift_system));
}