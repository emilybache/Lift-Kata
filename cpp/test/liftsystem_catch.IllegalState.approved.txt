  0   ]*A[    0
  1           1

